doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,282 @@
+may choose not to.
+
+If the runtime chooses to hide the window, there will still be a brief period during which that window is visible. It is
+inescapable.
+
+Likewise, any user running that GUI application from a console shell will see their shell hang until the application
+terminates.
+
+All of these scripting languages worked around this by shipping two binaries each, identical in every way expect in
+their subsystem fields. python/pythonw, perl/perlw, ruby/rubyw, wscript/cscript.
+
+PowerShell\[1\] is waiting to deal with this problem because they don't necessarily want to ship a `pwshw.exe` for all
+of their GUI-only authors. Every additional `*w` version of an application is an additional maintenance burden and
+source of cognitive overhead\[2\] for users.
+
+On the other side, you have mostly-GUI applications that want to print output to a console **if there is one
+connected**.
+
+These applications are still primarily GUI-driven, but they might support arguments like `/?` or `--help`. They only
+need a console when they need to print out some text. Sometimes they'll allocate their own console (which opens a new
+window) to display in, and sometimes they'll reattach to the originating console. VSCode does the latter, and so when
+you run `code` from CMD, and then `exit` CMD, your console window sticks around because VSCode is still attached to it.
+It will never print anything, and your only option is to close it.
+
+There's another risk risk in reattaching, too. Given that the shell decides whether to wait based on the subsystem
+field, GUI subsystem applications that reattach to their owning consoles *just to print some text* end up stomping on
+the output of any shell that doesn't wait for them:
+
+```
+C:\> application --help
+
+application - the interesting application
+C:\> Usage: application [OPTIONS] ...
+```
+
+> _(the prompt is interleaved with the output)_
+
+## Solution Design
+
+I propose that we introduce a fusion manifest field, **consoleWindowPolicy**, with the following values:
+
+* _absent_
+* `hidden`
+* `detached`
+
+These flags provide a way for an application to specify a default value for the console-related [process creation flags]
+supported by [`CreateProcess`] which will take effect when none are provided and there is no existing console session.
+
+It would look (roughly) like this:
+
+```xml
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
+  <application>
+    <windowsSettings>
+      <consoleWindowPolicy xmlns="http://schemas.microsoft.com/SMI/20XX/WindowsSettings">hidden</consoleWindowPolicy>
+    </windowsSettings>
+  </application>
+</assembly>
+```
+
+The effects of this field will only apply to binaries in the `IMAGE_SUBSYSTEM_WINDOWS_CUI` subsystem, as it pertains to
+the particulars of their console allocation.

```

Comment:
I feel like `consoleWindowPolicy` of `detached` should also be supported for binaries in the `IMAGE_SUBSYSTEM_WINDOWS_GUI` subsystem, which would signal that a console shell should wait for the process to exit, since older **Windows** versions don’t understand `consoleWindowPolicy`, so it’d be a possible breaking change for GUI apps that also log to a console if one is attached to switch to `IMAGE_SUBSYSTEM_WINDOWS_CUI` with `consoleWindowPolicy: detached`, since that would cause the GUI application to always create a console window in older **Windows** versions.

---

That should also avoid the concerns raised in <https://github.com/microsoft/terminal/pull/7337#discussion_r477282437>.

Reply:
But the console window policy for `...GUI` applications _is_ `detached`, and that's what this spec is attempting to remediate!

Reply:
Technically, it’s `detached` and don’t wait for the application to exit.

---

Because of this, if a `IMAGE_SUBSYSTEM_WINDOWS_GUI` application prints to the console, [then it and the shell will clobber each other’s output](https://github.com/microsoft/terminal/blob/main/doc/specs/%23607%20-%20Commandline%20Arguments%20for%20the%20Windows%20Terminal.md#subsystemwindows-or-subsystemconsole).

Reply:
Yes. There will be no changes to how shells wait based on this manifest field, because (1) they can’t read it and (2) I don’t expect every shell that exists to crack open an EXE to read its manifest.